MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetLoader.cs

```diff
@@ -146,35 +147,60 @@ private async Task LoadAssets(LoadAssets payload, Action onCompleteCallback)
 
 		private async Task<IList<Asset>> LoadAssetsFromGLTF(AssetSource source, Guid containerId, ColliderType colliderType)
 		{
-			IList<Asset> assets = new List<Asset>();
-			DeterministicGuids guidGenerator = new DeterministicGuids(UtilMethods.StringToGuid(
-				$"{containerId}:{source.ParsedUri.AbsoluteUri}"));
-
-			// download file
-			var rootUrl = URIHelper.GetDirectoryName(source.ParsedUri.AbsoluteUri);
-			var loader = new WebRequestLoader(rootUrl);
-			var stream = await loader.LoadStreamAsync(URIHelper.GetFileFromUri(source.ParsedUri));
+			// TODO: Cache the whole GLTF...
 
-			// pre-parse glTF document so we can get a scene count
-			// TODO: run this in thread
-			GLTF.GLTFParser.ParseJson(stream, out GLTF.Schema.GLTFRoot gltfRoot);
-			stream.Position = 0;
+			if (MREAPI.AppsAPI.VerboseLogging)
+			{
+				MREAPI.Logger.LogDebug($"Loading {source.ParsedUri.AbsoluteUri} for container {containerId}");
+			}
 
-			GLTFSceneImporter importer =
-				MREAPI.AppsAPI.GLTFImporterFactory.CreateImporter(gltfRoot, loader, _asyncHelper, stream);
-			importer.SceneParent = MREAPI.AppsAPI.AssetCache.CacheRootGO().transform;
-			importer.Collider = colliderType.ToGLTFColliderType();
+			IList<Asset> assets = new List<Asset>();
+			DeterministicGuids guidGenerator = new DeterministicGuids(UtilMethods.StringToGuid(
+				$"{source.ParsedUri.AbsoluteUri}"));
+			// $"{containerId}:{source.ParsedUri.AbsoluteUri}"));
+
+			Guid gltfid = guidGenerator.Next();
+			GLTFSceneImporter importer = null;
+			GLTF.Schema.GLTFRoot gltfRoot = null;
+#if CACHE_IMPORTERS
+			MREAPI.AppsAPI.AssetCache.TryGetGLTFImporter(gltfid, out importer, out gltfRoot);
+#endif
+			if (importer == null)
+			{
+				// download file
+				var rootUrl = URIHelper.GetDirectoryName(source.ParsedUri.AbsoluteUri);
+				var loader = new WebRequestLoader(rootUrl);
+				var stream = await loader.LoadStreamAsync(URIHelper.GetFileFromUri(source.ParsedUri));
+
+				// pre-parse glTF document so we can get a scene count
+				// TODO: run this in thread
+				GLTF.GLTFParser.ParseJson(stream, out gltfRoot);
+				stream.Position = 0;
+
+				importer = MREAPI.AppsAPI.GLTFImporterFactory.CreateImporter(gltfRoot, loader, _asyncHelper, stream);
+				importer.SceneParent = MREAPI.AppsAPI.AssetCache.CacheRootGO().transform;
+				importer.Collider = colliderType.ToGLTFColliderType();
+#if CACHE_IMPORTERS
+				MREAPI.AppsAPI.AssetCache.CacheGLTFImporter(gltfid, importer, gltfRoot);
+#endif
+			}
 
 			// load textures
 			if (gltfRoot.Textures != null)
 			{
 				for (var i = 0; i < gltfRoot.Textures.Count; i++)
 				{
-					await importer.LoadTextureAsync(gltfRoot.Textures[i], i, true);
-					var texture = importer.GetTexture(i);
-					texture.name = gltfRoot.Textures[i].Name ?? $"texture:{i}";
+					Guid id = guidGenerator.Next();
+
+					UnityEngine.Texture texture = MREAPI.AppsAPI.AssetCache.GetAsset(id) as UnityEngine.Texture;
+					if (texture == null)
```

Comment:
Ideally we'd never get this far. If assets already exist with the given ID, we should use those instead.